src/Features/CSharp/Portable/RemoveUnusedParametersAndValues/CSharpRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnusedParametersAndValues
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpRemoveUnusedParametersAndValuesDiagnosticAnalyzer : AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer
+    {
+        protected override Option<CodeStyleOption<UnusedValuePreference>> UnusedValueExpressionStatementOption
+            => CSharpCodeStyleOptions.UnusedValueExpressionStatement;
+
+        protected override Option<CodeStyleOption<UnusedValuePreference>> UnusedValueAssignmentOption
+            => CSharpCodeStyleOptions.UnusedValueAssignment;
+
+        protected override bool SupportsDiscard(SyntaxTree tree)
+            => ((CSharpParseOptions)tree.Options).LanguageVersion >= LanguageVersion.CSharp7;
+
+        protected override Location GetDefinitionLocationToFade(IOperation unusedDefinition)
+        {
+            switch (unusedDefinition.Syntax)
+            {
+                case VariableDeclaratorSyntax variableDeclartor:
+                    return variableDeclartor.Identifier.GetLocation();
+
+                case DeclarationPatternSyntax declarationPattern:
+                    return declarationPattern.Designation.GetLocation();
+
+                default:
+                    if (unusedDefinition.Syntax.Parent is ForEachStatementSyntax forEachStatement &&
+                        forEachStatement.Type == unusedDefinition.Syntax)
```

Comment:
that the unused definition is the 'foreach type' seems a bit strange to me...

Reply:
Is that an IOperation quirk?

Reply:
> Is that an IOperation quirk?

Precisely. C# syntax node for foreach statement has no syntax node for the loop control variable declaration, so we generate an `IVariableDeclarationOperation` with the syntax mapped to the type node instead of variable declaration, see [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.IOperation.UnitTests/IOperation/IOperationTests_IForEachLoopStatement.cs,35) for an example. I will add a comment clarifying this.

---
In reply to: [234795584](https://github.com/dotnet/roslyn/pull/30777#discussion_r234795584) [](ancestors = 234795584)

Reply:
Added comment

---
In reply to: [234795477](https://github.com/dotnet/roslyn/pull/30777#discussion_r234795477) [](ancestors = 234795477)

Reply:
Thanks!